### ISSUE TITLE
Refactor form-content overlap

### ISSUE THREAD

**Issue by agoose77** [scikit-hep/awkward#1738]
### Version of Awkward Array

main

### Description and code to reproduce

The `Form` class and `Content` class are highly related. Currently, the `Content` classes effectively use the `Form` class as a mixin, albeit via an indirect lookup:

```python
class Content:
    ...
    
    @property
    def purelist_isregular(self) -> bool:
        return self.Form.purelist_isregular.__get__(self)
```

The `Form` and `Content` classes look very similar; they share a common set of attributes, in addition to similarly-named attributes that have different types.

It would be better (and slightly faster, although this is not a performance argument) if we implemented this as a true mixin class, e.g `FormLike` or `FormMixin`. Not only do we get rid of a set of indirect calls via `@property` decorators, we also decouple the forms from the contents through a third class.

Although composition is an option here - i.e. introducing a `Metadata` class that is owned by both `Form` and `Content`, it would probably complicate things more than help, as all three classes would need to set-up a tree structure. With mixins, we only need two classes to do this (`Form`, `Content`).

**Comment by jpivarski** [comment]
I'm not so sure about this being an improvement (again, ignoring the performance argument). All of the alternatives to what we have now introduce new classes, whether by composition or containment. Classes draw a lot of attention when a contributor or maintainer is looking at the code, and it might be drowning out attention that should be spent elsewhere. There needs to be a relatively high bar on what deserves to be a class (at least, for classes that must be visible at the top level of a project). Do we really want `Content`, `Form`, `Type`, and `Metadata` (or `FormMixin`) with a composition or containment relationship between the fourth and the first two? Does the fourth need to be specialized for List, ListOffset, BitMasked, etc.? Because that could get complicated.

The way it is now, it's following a rule. Some methods/properties are needed on both a `Content` and its corresponding `Form`. We also don't want to have the `Content` create a `Form` just to call this one method on it (performance). The rule is that such a method/property should be implemented on the `Form` in a duck-typeable way and then called by the corresponding `Content` (not the other way around). Except for the fact that properties need to be called with an ugly `.__get__(self)`, I think that's a simple rule and it doesn't call too much attention to itself. It's best to keep these methods/properties next to each other in the source code so that they can be thought of as a group, and hopefully anyone reading the code would catch onto the pattern (maybe comments or docstrings would help).

-------------

For an example of class inheritance (that I wrote, on purpose) that looks overwhelming when you sit back and look at it, this happened in Vector (diagram indicates both composition and containment, in different colors):

<img src="https://raw.githubusercontent.com/scikit-hep/vector/main/docs/api/inheritance.svg" width="100%">

We had a bunch of constraints, such as not reimplementing 2D methods/properties on 3D and 4D vectors, having multiple backends, a distinction between pure geometry vectors and momentum vectors, etc. The constraints could have been implemented in highly dynamic code, in which concrete classes for each case are generated by `exec` from strings or using the `type` constructor (or in languages that have them, macros), or the constraints could have been implemented statically using predefined classes with OOP, which is what that diagram represents. The latter also allows for static typing, another constraint, so that's why we chose this method and don't regret it. My point is just, "Look what happens when complicated things are expressed statically, with composition and containment. It's possible, but look at it!"

Applying it to the case in hand, having a `Content` subclass call its corresponding `Form` subclass's method/property as if it were its own is a dynamic "cheat." The pejorative word "cheat" only applies if you're committed to orthodox OOP. To use another pejorative word, it's "lying" about what "`self`" means: a `Form` subclass is getting passed a `self` variable that is not a member of its class. It works because Python is dynamic: methods are just functions, types are just objects, and everything is a `PyObject*` that can be passed to any function. There are static, OOP ways of doing it, but they increase the number of classes and have non-trivial relationships with each other.

Is this extra complexity, purchased in exchange for adhering to a static OOP principle, worth it in this case? I didn't think so when I wrote it, but if you feel strongly about it, I'll reconsider.

**Comment by agoose77** [comment]
I'm not wedded to "pure" OOP (in fact, I find oftentimes class-based implementations are writen without good reason). You've already outlined some of the strategies available for vector, and I'm sure that there are others that we haven't considered yet. I wasn't involved with that design phase, so I won't try and think of any new ones off the top of my head ;) I agree that in vector, this all gets very complicated. But, vector is also doing quite complicated things, certainly compared to what we expect to see in Awkward.

I think it's worth noting that there are two axes here: 
- implementation sharing
- type hinting

My motivation/goals are
- decouple `Form` and `Content`: they are closely related, but `Form` shouldn't need to know about `Content`. Right now, although it's a limited API surface, forms do need to know about contents - we call `Form.method.__get__(content)`. Right now we have this rigid coupling requirement, but it's not enforced by the type system.
- creating an inheritance tree makes the relationship more readable, i.e. `Form` and `Content` are both specialisms of a more general array metadata system. I was originally pro-composition here, but actually looking at this quickly made it clear that `Form` and `Content` are more like sibling types than "owner-property".
- type checking (self-explanatory)


Rather than making this argument with words though, I wrote up a small example that compares these two implementations:

<details><summary>Current</summary>

```python
import typing


class Form:
    @property
    def purelist_depth(self) -> int:
        raise NotImplementedError


class Content:
    @property
    def form(self) -> Form:
        raise NotImplementedError

    @property
    def Form(self) -> type[Form]:
        raise NotImplementedError

    @property
    def purelist_depth(self) -> int:
        return typing.cast(int, self.Form.purelist_depth.__get__(self))


# List ##############################
class ListForm(Form):
    def __init__(self, content: Form):
        self.content = content

    @property
    def purelist_depth(self) -> int:
        return self.content.purelist_depth + 1


class ListContent(Content):
    def __init__(self, content: Content):
        self.content = content

    @property
    def form(self) -> ListForm:
        return ListForm(self.content.form)

    Form = ListForm


# Numpy ##############################
class NumpyForm(Form):
    @property
    def purelist_depth(self) -> int:
        return 1


class NumpyContent(Content):
    @property
    def form(self) -> NumpyForm:
        return NumpyForm()

    Form = NumpyForm


# Test ###############################
array = ListContent(ListContent(NumpyContent()))


def get_array_depth(array: Content | Form) -> int:
    return array.purelist_depth


print(
    get_array_depth(array)
)
print(
    get_array_depth(array.form)
)
```

</details>

<details><summary>Mixins</summary>

```python
class ContentLike:
    @property
    def purelist_depth(self) -> int:
        raise NotImplementedError


class Form(ContentLike):

    ...


class Content(ContentLike):
    @property
    def form(self) -> Form:
        raise NotImplementedError


# List ##############################
class ListLike(ContentLike):

    content: ContentLike

    @property
    def purelist_depth(self) -> int:
        return self.content.purelist_depth + 1


class ListForm(ListLike, Form):

    content: Form

    def __init__(self, content: Form):
        self.content = content


class ListContent(ListLike, Content):

    content: Content

    def __init__(self, content: Content):
        self.content = content

    @property
    def form(self) -> ListForm:
        return ListForm(self.content.form)


# Numpy ##############################
class NumpyLike:
    @property
    def purelist_depth(self) -> int:
        return 1


class NumpyForm(NumpyLike, Form):
    ...


class NumpyContent(NumpyLike, Content):
    @property
    def form(self) -> NumpyForm:
        return NumpyForm()


# Test ###############################
array = ListContent(ListContent(NumpyContent()))


def get_array_depth(array: ContentLike) -> int:
    return array.purelist_depth


print(
    get_array_depth(array)
)
print(
    get_array_depth(array.form)
)
```

</details>

This is not a pure mixins approach - we have diamond inheritance here. But, Python's smart enough to resolve this, and it means we can ask for `ListLike` as an interface, instead of requiring an intersection type that I don't think Python's typing system supports yet.[^protocol]

For the sake of a new class per content-type, I don't think this is all that much more complexity.

[^protocol]: you could create a new intersection protocol to do this, but that's just more code for the sake of typing.

**Comment by agoose77** [comment]
Closed by #2841 :)
